x-pack/plugin/apm-integration/src/main/java/org/elasticsearch/xpack/apm/APMTracer.java

```diff
@@ -111,20 +110,37 @@ protected void doStop() {
     protected void doClose() {}
 
     @Override
-    public void onTaskRegistered(Task task) {
+    public void onTraceStarted(TracingPlugin.Traceable traceable) {
         final Tracer tracer = this.tracer;
         if (tracer != null) {
-            taskSpans.computeIfAbsent(task.getId(), taskId -> {
-                final Span span = tracer.spanBuilder(task.getAction()).startSpan();
-                span.setAttribute("es.task.id", task.getId());
+            spans.computeIfAbsent(traceable.getSpanId(), spanId -> {
+                final Span span = tracer.spanBuilder(traceable.getSpanName()).startSpan();
```

Comment:
When setting attributes while creating the span, it's common to use `SpanBuilder.setAttribute(key, value)`. It would look like:

```java
SpanBuilder spanBuilder = tracer.spanBuilder(traceable.getSpanName());
for (Map.Entry<String, Object> entry : traceable.getAttributes().entrySet()) {
    Object value = entry.getValue();
    if (value instanceof String) {
        spanBuilder.setAttribute(entry.getKey(), (String) value);
    } else if (value instanceof Long) {
        spanBuilder.setAttribute(entry.getKey(), (Long) value);
    } else if (value instanceof Integer) {
        spanBuilder.setAttribute(entry.getKey(), (Integer) value);
    } else if (value instanceof Double) {
        spanBuilder.setAttribute(entry.getKey(), (Double) value);
    } else if (value instanceof Boolean) {
        spanBuilder.setAttribute(entry.getKey(), (Boolean) value);
    } else {
        throw new IllegalArgumentException(
            "span attributes do not support value type of [" + value.getClass().getCanonicalName() + "]"
    );
}
return spanBuilder.startSpan();
```

Reply:
Thanks I updated as suggested.